Automatically close on session.reset()

pontusmelke opened the pull request:
In the event of a reset the transaction should be completely closed and the
session should be ready to create new transactions without explicitly closing
any transactions.

pontusmelke committed:
Automatically close on session.reset()

In the event of a reset the transaction should be completely closed and the
session should be ready to create new transactions without explicitly closing
any transactions.

zhenlineo commented:
So the synchronized modifier is to enforce changes to `currentTx` in `session` are all sequential. I can hardly think of any simpler way to make it more thread-safe.

Commit:
Made reset method sync so that once `isInterrupted` atomic filed get set, no more messages will be enqueued or flushed until the current sent ones has been synced.
Always ackFail

Commit:
Muted ack_failure while a reset is sent and no reply has been recived

Given the following message senquence between server and client

C           S
RUN      -> FAIL, error e
PULL_ALL -> IGNORED
RESET    -> SUCC

In the previous PR, the error e could be thrown either after RESET or before RESET depending on if we called sync (receiveAll) after RESET or before RESET.
In this PR, by replacing receiveAll with receiveOne multiple times, we ensure that if the sesssion get interrupted, the error e is always immediately thrown before reset. Therefore we know we could mute akc_failure after a reset is sent, before a reply to this reset is received.
